crates/solidity/inputs/language/bindings/rules.msgb

```diff
@@ -24,11 +25,39 @@ inherit .enclosing_def
   ;; This provides all the exported symbols from the file
   node @source_unit.defs
 
+  ;; Connect to ROOT_NODE to export our symbols
   node export
   attr (export) pop_symbol = FILE_PATH
   edge ROOT_NODE -> export
   edge export -> @source_unit.defs
 
+  ;; Automatically import language built-ins
+  if (not (eq FILE_PATH BUILT_INS_FILE_PATH)) {
+    ;; All built-in symbols are defined in an internal contract named '$$'
+    ;; so we need to construct an equivalent import path to reach them.
+    ;; We should have access to both type members (eg. defined enums & structs)
+    ;; as well as functions and state variables (see special case below), hence
+    ;; why we're introducing a path through `@typeof`.
+    node built_ins
+    attr (built_ins) push_symbol = BUILT_INS_FILE_PATH
+
+    node built_in_library
+    attr (built_in_library) push_symbol = "$$"
+
+    node built_in_typeof
+    attr (built_in_typeof) push_symbol = "@typeof"
+
+    node built_in_member
+    attr (built_in_member) push_symbol = "."
+
+    edge @source_unit.lexical_scope -> built_in_member
```

Comment:
I think I'm missing something here:

1. Why would we use push symbols (`.`, `@typeof`, etc..) here? the users can never write something like `$BuiltIns$.XXX` anyways. Can't we just link the lexical scope of the `ContractMembers` to the parent file scope?
2. What here prevents users from referencing the `$BuiltIns$` contract directly in their code? and if they can't, what is the significance of using dollar signs in the first place? why can't we just use `BuiltIns` for example?

Happy to follow up offline to understand better. Not a blocker.

Reply:
1. Since built-ins are inside the special `$BuiltIns$` contract, we need a way to "promote" them to the global namespace. That's why `.` and `@typeof` are pushed here, in connection to the built-ins path. This is _as if_ the user had referenced global with the `$BuiltIns$.` prefix.

Having said that, I think we can find a better way to do this, by directly linking the `$BuiltIns$` members to the global namespace _from the built-ins side_.

2. For now, nothing prevents users from referencing `$BuiltIns$`. I have some code in another branch that renames all identifiers in the built-ins file by replacing `$` for `%`. That way it will be impossible for users to explicitly reference the built-ins contract. Similarly with built-in types such as `address`.

3. That's a good point. For types it shouldn't be a problem because of the renaming of built-in identifiers. But we currently don't have any way to prefer a user's definition over the built-in. We could resolve it via ranking, but since we've discussed about removing that feature, we may need to fallback to a nano pass later.

Reply:
> I have some code in another branch that renames all identifiers in the built-ins file by replacing $ for %.

Instead of that, similar to how we link its inner members to the global scope, can we just exclude the parent from the global scope (if it has that name, and in a system file)?
Therefore, we won't need any special processing, and it won't conflict with any user-defined types with the same name.

Reply:
> We could resolve it via ranking, but since we've discussed about removing that feature, we may need to fallback to a nano pass later.

can it be just solved via stack graph precedence? give the `import "builtins.sol"` edge the lowest precedence possible, so that we try to bind to the user-defined values first?
I'm thinking of what happens in the scenario of:

```solidity
import "libA.sol"; // exposes a "Foo" type
import "libB.sol"; // exposes a "Foo" type as well

Foo foo; // this
```

In that case, Solidity expects to bind `Foo` from `libA`, and declares any other conflicting symbols from the later import `libB` an error/duplicate definition.
Do we use precedence here as well?

> For types it shouldn't be a problem because of the renaming of built-in identifiers

That's the part I don't understand. We define types like `$BuiltIn$TxType` but rename it to `TxType` at some point?
Not sure why we do that, if precedence above works as expected. Why can't we just use `TxType` normally?

But of course not a blocker. I'm happy to follow up on this to understand better.

Reply:
> Instead of that, similar to how we link its inner members to the global scope, can we just exclude the parent from the global scope (if it has that name, and in a system file)?
> Therefore, we won't need any special processing, and it won't conflict with any user-defined types with the same name.

I'm not sure I follow here. We would not be able to exclude the type and link directly to their members because then those members would be available globally and/or the resolution when doing member access would not work.

We do need some kind of renaming because it's not possible to use some of the built-in types as identifiers, because they are keywords in the grammar (eg. `address`). So we wouldn't be able to define the members of address with normal parsing.

The binding rules expose the built-in types by prefixing their names with a `%` symbol, and pushing the result to the symbol stack. That way we can "define" the built-in members by parsing, eg. `struct %address { ... }`. Of course, since `%` is not a valid identifier, we'll use `$` for rendering the built-ins and rename the identifier by post-processing the CST tree.

Reply:
> In that case, Solidity expects to bind Foo from libA, and declares any other conflicting symbols from the later import libB an error/duplicate definition.
Do we use precedence here as well?

There is nothing in place to handle that, and I didn't have it in the radar. My gut feeling is that this should be handled later, not shadowed by a precedence attribute. A compiler will want to report this accidental shadowing, and if we handle it at the stack graph it will not be immediate to report. Or at least it will fall in the same category as a variable shadowed by scope (although I'm not sure now if we wouldn't like to report that as well).

Reply:
>We define types like $BuiltIn$TxType but rename it to TxType at some point?
Not sure why we do that, if precedence above works as expected. Why can't we just use TxType normally?

Those types are only declared to represent the global variables and their members. The name I settled on is [`$txType`](https://github.com/manastech/slang/blob/33ecc0552be48ef5040733dfc8f1df312c01faae/crates/solidity/outputs/cargo/slang_solidity/src/generated/bindings/generated/built_ins/0.8.26.sol#L70) in this case, but this [gets renamed](https://github.com/manastech/slang/blob/33ecc0552be48ef5040733dfc8f1df312c01faae/crates/solidity/outputs/cargo/slang_solidity/src/lib.rs#L6-L45) to `%txType`. The user can still redefine `tx` as a variable, but will not be able to redefine the internal `%txType`.

Reply:
> So we wouldn't be able to define the members of address with normal parsing.

That might be worth fixing in the grammar as well. Maybe `address` should be unreserved here? I will look into it separetely.
Is there other corner cases like `address`? I'm trying to see if I can help simplifying this workflow.